aws/resource_aws_traffic_mirror_session.go

```diff
@@ -0,0 +1,212 @@
+		input.Description = aws.String(v.(string))
+	}
+
+	if v, ok := d.GetOk("packet_length"); ok {
+		input.PacketLength = aws.Int64(int64(v.(int)))
+	}
+
+	if v, ok := d.GetOk("session_number"); ok {
+		input.SessionNumber = aws.Int64(int64(v.(int)))
+	}
+
+	if v, ok := d.GetOk("virtual_network_id"); ok {
+		input.VirtualNetworkId = aws.Int64(int64(v.(int)))
+	}
+
+	out, err := conn.CreateTrafficMirrorSession(input)
+	if nil != err {
+		return fmt.Errorf("Error creating traffic mirror session %v", err)
+	}
+
+	d.SetId(*out.TrafficMirrorSession.TrafficMirrorSessionId)
+	return resourceAwsTrafficMirrorSessionRead(d, meta)
+}
+
+func resourceAwsTrafficMirrorSessionUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	sessionId := d.Id()
+	input := &ec2.ModifyTrafficMirrorSessionInput{
+		TrafficMirrorSessionId: &sessionId,
+	}
+
+	if d.HasChange("session_number") {
+		_, n := d.GetChange("session_number")
+		input.SessionNumber = aws.Int64(int64(n.(int)))
+	}
+
+	if d.HasChange("traffic_mirror_filter_id") {
+		_, n := d.GetChange("traffic_mirror_filter_id")
+		input.TrafficMirrorFilterId = aws.String(n.(string))
+	}
+
+	if d.HasChange("traffic_mirror_target_id") {
+		_, n := d.GetChange("traffic_mirror_target_id")
+		input.TrafficMirrorTargetId = aws.String(n.(string))
+	}
+
+	var removeFields []*string
+	if d.HasChange("description") {
+		_, n := d.GetChange("description")
+		if "" != n {
+			input.Description = aws.String(n.(string))
+		} else {
+			removeFields = append(removeFields, aws.String("description"))
+		}
+	}
+
+	if d.HasChange("packet_length") {
+		_, n := d.GetChange("packet_length")
+		if nil != n && n.(int) > 0 {
+			input.PacketLength = aws.Int64(int64(n.(int)))
+		} else {
+			removeFields = append(removeFields, aws.String("packet-length"))
+		}
+	}
+	log.Printf("[DEBUG] removeFields %v", removeFields)
+
+	if d.HasChange("virtual_network_id") {
+		_, n := d.GetChange("virtual_network_id")
+		log.Printf("[DEBUG] VNI has change %v", n)
+		if nil != n && n.(int) > 0 {
+			input.VirtualNetworkId = aws.Int64(int64(n.(int)))
+		} else {
+			removeFields = append(removeFields, aws.String("virtual-network-id"))
+		}
+	}
+
+	log.Printf("[DEBUG] removeFields %v", removeFields)
+	if len(removeFields) > 0 {
+		input.SetRemoveFields(removeFields)
+	}
+
+	_, err := conn.ModifyTrafficMirrorSession(input)
+	if nil != err {
+		return fmt.Errorf("Error updating traffic mirror session %v", err)
+	}
+
+	return resourceAwsTrafficMirrorSessionRead(d, meta)
+}
+
+func resourceAwsTrafficMirrorSessionRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).ec2conn
+
+	sessionId := d.Id()
+	input := &ec2.DescribeTrafficMirrorSessionsInput{
+		TrafficMirrorSessionIds: []*string{
+			&sessionId,
+		},
+	}
+
+	out, err := conn.DescribeTrafficMirrorSessions(input)
+	if nil != err {
+		d.SetId("")
+		return fmt.Errorf("Error describing traffic mirror session %v", sessionId)
+	}
```

Comment:
We should not remove this resource from the Terraform state on API errors and should return the error itself.

```
if err != nil {
	return fmt.Errorf("error describing EC2 Traffic Mirror Session (%s): %w", d.Id(), err)
}
```